tests/src/CancellationToken.cpp

```diff
@@ -0,0 +1,169 @@
+#include "util/CancellationToken.hpp"
+
+#include <gtest/gtest.h>
+
+using namespace chatterino;
+
+// CancellationToken
+
+TEST(CancellationToken, ctor)
+{
+    {
+        CancellationToken token;
+        ASSERT_TRUE(token.isCancelled());
+        token.cancel();
+        ASSERT_TRUE(token.isCancelled());
+    }
+    {
+        CancellationToken token(false);
+        ASSERT_FALSE(token.isCancelled());
+        token.cancel();
+        ASSERT_TRUE(token.isCancelled());
+    }
+    {
+        CancellationToken token(true);
+        ASSERT_TRUE(token.isCancelled());
+        token.cancel();
+        ASSERT_TRUE(token.isCancelled());
+    }
+}
+
+TEST(CancellationToken, moveCtor)
+{
+    CancellationToken token(false);
+    ASSERT_FALSE(token.isCancelled());
+    CancellationToken token2(std::move(token));
+    ASSERT_TRUE(token.isCancelled());
+    ASSERT_FALSE(token2.isCancelled());
+
+    token.cancel();
+    ASSERT_FALSE(token2.isCancelled());
+
+    token2.cancel();
+    ASSERT_TRUE(token2.isCancelled());
+}
+
+TEST(CancellationToken, moveAssign)
+{
+    CancellationToken token(false);
+    CancellationToken token2;
+    ASSERT_FALSE(token.isCancelled());
+    ASSERT_TRUE(token2.isCancelled());
+
+    token2 = std::move(token);
+    ASSERT_TRUE(token.isCancelled());
+    ASSERT_FALSE(token2.isCancelled());
+    token.cancel();
+    ASSERT_FALSE(token2.isCancelled());
+
+    token2.cancel();
+    ASSERT_TRUE(token2.isCancelled());
+}
+
+TEST(CancellationToken, copyCtor)
+{
+    CancellationToken token(false);
+    ASSERT_FALSE(token.isCancelled());
+    CancellationToken token2(token);
+    ASSERT_FALSE(token.isCancelled());
+    ASSERT_FALSE(token2.isCancelled());
+    token2.cancel();
+    ASSERT_TRUE(token2.isCancelled());
+    ASSERT_TRUE(token.isCancelled());
+}
+
+TEST(CancellationToken, copyAssign)
+{
+    CancellationToken token(false);
+    CancellationToken token2;
+    ASSERT_FALSE(token.isCancelled());
+    ASSERT_TRUE(token2.isCancelled());
+
+    token2 = token;
+    ASSERT_FALSE(token.isCancelled());
+    ASSERT_FALSE(token2.isCancelled());
+    token2.cancel();
+    ASSERT_TRUE(token.isCancelled());
+    ASSERT_TRUE(token2.isCancelled());
+}
+
+TEST(CancellationToken, dtor)
+{
+    CancellationToken token(false);
+    {
+        // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+        CancellationToken token2 = token;
+        ASSERT_FALSE(token.isCancelled());
+        ASSERT_FALSE(token2.isCancelled());
+    }
+
+    ASSERT_FALSE(token.isCancelled());
+}
+
+// ScopedCancellationToken
+
+TEST(ScopedCancellationToken, moveCancelCtor)
+{
+    CancellationToken token(false);
+    ASSERT_FALSE(token.isCancelled());
+    {
+        ScopedCancellationToken scoped(std::move(token));
```

Comment:
Correct me if I'm wrong:
Since `ScopedCancellationToken` has no accessors on its own, moving a `CancellationToken` into a `ScopedCancellationToken` means no one can check its state, I feel like unless `ScopedCancellationToken` has its own accessors for its `backingToken_`, the move ctor & move assignment operator should be deleted

Reply:
It takes the `CancellationToken` by value now. But generally, the idea is that using a move, you can avoid one copy of the `shared_ptr`. It's not really visible in the constructor as much but more in the assignment operator:

```cpp
CancellationToken token(false);
sendRequest("/foo", token /* copied */);
this->myToken = std::move(token); // avoid any copy of the shared_ptr
```